checks/evaluation/permissions.go

```diff
@@ -0,0 +1,315 @@
+	"github.com/ossf/scorecard/v4/probes/hasGitHubWorkflowPermissionNone"
+	"github.com/ossf/scorecard/v4/probes/hasGitHubWorkflowPermissionRead"
+	"github.com/ossf/scorecard/v4/probes/hasGitHubWorkflowPermissionUndeclared"
+	"github.com/ossf/scorecard/v4/probes/hasGitHubWorkflowPermissionUnknown"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteActionsJob"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteActionsTop"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteAllJob"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteAllTop"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteChecksJob"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteChecksTop"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteContentsJob"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteContentsTop"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteDeploymentsJob"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteDeploymentsTop"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWritePackagesJob"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWritePackagesTop"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteSecurityEventsJob"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteSecurityEventsTop"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteStatusesJob"
+	"github.com/ossf/scorecard/v4/probes/hasNoGitHubWorkflowPermissionWriteStatusesTop"
+)
+
+// TokenPermissions applies the score policy for the Token-Permissions check.
+func TokenPermissions(name string,
+	findings []finding.Finding,
+	dl checker.DetailLogger,
+) checker.CheckResult {
+	expectedProbes := []string{
+		hasNoGitHubWorkflowPermissionWriteActionsTop.Probe,
+		hasNoGitHubWorkflowPermissionWriteAllTop.Probe,
+		hasNoGitHubWorkflowPermissionWriteChecksTop.Probe,
+		hasNoGitHubWorkflowPermissionWriteContentsTop.Probe,
+		hasNoGitHubWorkflowPermissionWriteDeploymentsTop.Probe,
+		hasNoGitHubWorkflowPermissionWritePackagesTop.Probe,
+		hasNoGitHubWorkflowPermissionWriteSecurityEventsTop.Probe,
+		hasNoGitHubWorkflowPermissionWriteStatusesTop.Probe,
+		hasGitHubWorkflowPermissionUnknown.Probe,
+		hasGitHubWorkflowPermissionNone.Probe,
+		hasGitHubWorkflowPermissionRead.Probe,
+		hasGitHubWorkflowPermissionUndeclared.Probe,
+		hasNoGitHubWorkflowPermissionWriteAllJob.Probe,
+		hasNoGitHubWorkflowPermissionWriteSecurityEventsJob.Probe,
+		hasNoGitHubWorkflowPermissionWriteActionsJob.Probe,
+		hasNoGitHubWorkflowPermissionWriteContentsJob.Probe,
+		hasNoGitHubWorkflowPermissionWritePackagesJob.Probe,
+		hasNoGitHubWorkflowPermissionWriteChecksJob.Probe,
+		hasNoGitHubWorkflowPermissionWriteDeploymentsJob.Probe,
+		hasNoGitHubWorkflowPermissionWriteStatusesJob.Probe,
+	}
```

Comment:
I only got a few lines in, until I got to here. I think this is an excessive amount of probes, and I'm almost done writing my reasoning (will be in its own issue). Without looking at the rest of the code, I assume the amount of probes is due to limitations around the Values map?

If the values map was changed to be `map[string]string`, do you imagine these could be condensed?

e.g. 

```go
f = f.WithValues(map[string]int{
	"Level": "read",
	"Permission": "contents",
	// and so on
})
```

Reply:
Yeah, I was thinking about a similar change as well.

Or, rather, I thought of `map[string]comparable`, which would allow each Probe to use whichever type is best for it's particular use-case:

```go
type Finding[T comparable] struct {
	// ...
	Values      map[string]T     `json:"values,omitempty"`
	// ...
}
```

(no idea how generics interact with the JSON annotation)

The generic `comparable` ensures the type is capable of `==`, `>`, `<`, etc. The check evaluations are almost all, at the end of the day, just a series of such comparisons, so this should work somewhat seamlessly. Any evaluations which require something more advanced would need to cast the values to the underlying type.

Reply:
You can't make a slice of []Finding anymore if some are Finding[string] and some are Finding[int].
https://go.dev/play/p/VlnPINUIFXh

Reply:
the extreme option is  `Values map[string]any`, but then you need to deal with type stuff anyway

Reply:
One thing we could do here is combine the top-level permissions into one probe and job-level permissions into another probe. 

However, as I see it, this way we might fall short when it comes to matching the outcome with the security impact. `OutcomePositive` will not always be positive but will just answer "yes", and vice versa for `OutcomeNegative`. For example, if we have a probe that is called `hasTopLevelPermissions` and create one finding per top level permission, then we will return `OutcomePositive` to answer the probes question, but in some cases the findings have negative security impact. The same goes - although less so - for job-level permissions: A probe that is called `hasJobLevelPermissions` should return `OutcomePositive` to answer `yes`, but if the token type is `all`, then we should return `OutcomeNegative`, however it has a negative security impact on the project.

This division might make sense:

| Probe | Outcome | Findings |
| ------------- | ------------- | ------------- |
| freeOfTopLevelWritePermissions  | Positive outcome if the project has no top level write permissions. Negative outcome if the project has top level write permissions. | The probe returns one finding per top-level write permission token found. The token type is specified with a value |
| freeOfJobLevelWritePermissions| Positive outcome if the project has no job level write permissions. Negative outcome if the project has job level write permissions. | The probe returns one finding per job-level write permission token found. The token type is specified with a value |

The first probe would cover these (current) probes:

- hasNoGitHubWorkflowPermissionWriteActionsTop.Probe,
- hasNoGitHubWorkflowPermissionWriteAllTop.Probe,
- hasNoGitHubWorkflowPermissionWriteChecksTop.Probe,
- hasNoGitHubWorkflowPermissionWriteContentsTop.Probe,
- hasNoGitHubWorkflowPermissionWriteDeploymentsTop.Probe,
- hasNoGitHubWorkflowPermissionWritePackagesTop.Probe,
- hasNoGitHubWorkflowPermissionWriteSecurityEventsTop.Probe,
- hasNoGitHubWorkflowPermissionWriteStatusesTop.Probe,

The second probe would cover these (current) probes:

- hasNoGitHubWorkflowPermissionWriteAllJob.Probe,
- hasNoGitHubWorkflowPermissionWriteSecurityEventsJob.Probe,
- hasNoGitHubWorkflowPermissionWriteActionsJob.Probe,
- hasNoGitHubWorkflowPermissionWriteContentsJob.Probe,
- hasNoGitHubWorkflowPermissionWritePackagesJob.Probe,
- hasNoGitHubWorkflowPermissionWriteChecksJob.Probe,
- hasNoGitHubWorkflowPermissionWriteDeploymentsJob.Probe,
- hasNoGitHubWorkflowPermissionWriteStatusesJob.Probe,

The remaining probes would stay the same.

The second probe would make it easy to log if there are no job-level permissions:

https://github.com/ossf/scorecard/blob/fcb1eed20bbe9bcb24d591a9aba84ef4a6a19828/checks/evaluation/permissions.go#L189-L204

However, the outcome of the second probe would not always answer "yes" + consider the security impact. For example, the evaluation only scores based on the `all` token type at the job level and logs the rest. As such, we could add a probe that is called `hasNoAllPermissionAtTopAndJobLevel` which eliminates a fair bit of processing in the evaluation. This would be a highly opinionated probe: `OutcomeNegative` = `minScore` from here:

https://github.com/ossf/scorecard/blob/fcb1eed20bbe9bcb24d591a9aba84ef4a6a19828/checks/evaluation/permissions.go#L142-L145

https://github.com/ossf/scorecard/blob/fcb1eed20bbe9bcb24d591a9aba84ef4a6a19828/checks/evaluation/permissions.go#L152-L155

This would also reduce this part to a simple check of the outcome from `hasNoAllPermissionAtTopAndJobLevel`: 

https://github.com/ossf/scorecard/blob/fcb1eed20bbe9bcb24d591a9aba84ef4a6a19828/checks/evaluation/permissions.go#L134-L158

Reply:
> One thing we could do here is combine the top-level permissions into one probe and job-level permissions into another probe.

I think this top-level vs job-level division makes sense.

> However, as I see it, this way we might fall short when it comes to matching the outcome with the security impact. OutcomePositive will not always be positive but will just answer "yes", and vice versa for OutcomeNegative

Yes, it can be confusing with the current Positive/Negative. For other situations such as commit activity, I've thought about having an `OutcomeInformational`, or similar, when we want to leave judgement up to the consumer.

Reply:
> Yes, it can be confusing with the current Positive/Negative. For other situations such as commit activity, I've thought about having an OutcomeInformational, or similar, when we want to leave judgement up to the consumer.

That would be useful in this case.

Reply:
Main comment issue was addressed, and follow-up outcome discussion should be discussed/handled  in #3866